shell/common/shell_unittests.cc

```diff
@@ -952,5 +953,69 @@ TEST_F(ShellTest, IsolateCanAccessPersistentIsolateData) {
   DestroyShell(std::move(shell), std::move(task_runners));
 }
 
+TEST_F(ShellTest, Screenshot) {
+  auto settings = CreateSettingsForFixture();
+  std::unique_ptr<Shell> shell = CreateShell(settings);
+
+  // Create the surface needed by rasterizer
+  PlatformViewNotifyCreated(shell.get());
+
+  auto configuration = RunConfiguration::InferFromSettings(settings);
+  configuration.SetEntrypoint("emptyMain");
+
+  RunEngine(shell.get(), std::move(configuration));
+
+  LayerTreeBuilder builder = [&](std::shared_ptr<ContainerLayer> root) {
+    SkPictureRecorder recorder;
+    SkCanvas* recording_canvas =
+        recorder.beginRecording(SkRect::MakeXYWH(0, 0, 80, 80));
+    recording_canvas->drawRect(SkRect::MakeXYWH(0, 0, 80, 80),
+                               SkPaint(SkColor4f::FromColor(SK_ColorRED)));
+    auto sk_picture = recorder.finishRecordingAsPicture();
+    fml::RefPtr<SkiaUnrefQueue> queue = fml::MakeRefCounted<SkiaUnrefQueue>(
+        this->GetCurrentTaskRunner(), fml::TimeDelta::FromSeconds(0));
+    auto picture_layer = std::make_shared<PictureLayer>(
+        SkPoint::Make(10, 10),
+        flutter::SkiaGPUObject<SkPicture>({sk_picture, queue}), false, false);
+    root->Add(picture_layer);
+  };
+
+  PumpOneFrame(shell.get(), 100, 100, builder);
+
+  fml::Status result =
+      shell->WaitForFirstFrame(fml::TimeDelta::FromMilliseconds(1000));
+  ASSERT_TRUE(result.ok());
+
+  std::shared_ptr<fml::AutoResetWaitableEvent> latch =
+      std::make_shared<fml::AutoResetWaitableEvent>();
+
+  Rasterizer::Screenshot screenshot;
+  fml::TaskRunner::RunNowOrPostTask(
+      shell->GetTaskRunners().GetGPUTaskRunner(), [&]() {
+        auto rasterizer = shell->GetRasterizer();
+        screenshot = rasterizer->ScreenshotLastLayerTree(
+            Rasterizer::ScreenshotType::CompressedImage, true);
+        latch->Signal();
+      });
+  latch->Wait();
+
+  const char* reference_png =
+      "iVBORw0KGgoAAAANSUhEUgAAAGQAAABkCAYAAABw4pVUAAAAAXNSR0IArs4c6QAAAARzQklU"
+      "CAgICHwIZIgAAADpSURBVHic7dHBCcAwDARBOf337FQQSB6GJcy89Ti0MwAAAEDHenu4Z/"
+      "bJIX+"
+      "3Xv76Oj2EbwSJESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgS"
+      "JESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgS"
+      "JESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgSJESRGkBhBYgS"
+      "JESRGkBhBYgSJESRGkBhBYgSJEQQAAADg0Q0wHwKgbFGDCQAAAABJRU5ErkJggg==";
+  // Use MakeWithoutCopy instead of MakeWithCString because we don't want to
+  // encode the null sentinel
+  sk_sp<SkData> reference_data =
+      SkData::MakeWithoutCopy(reference_png, strlen(reference_png));
+
+  ASSERT_TRUE(reference_data->equals(screenshot.data.get()));
```

Comment:
Skia in different OSes may generate different renderings. Before we can use Skia gold in the engine, maybe we need to conservatively [only compare golden images on Linux](https://github.com/flutter/engine/blob/master/flow/layers/performance_overlay_layer_unittests.cc#L81).

Reply:
Let's avoid this line of reasoning.The engine is built and tested on multiple platforms and we still allow developers to have Macs as their primary development environment. The embedder unit-test target has been using Skia with SwiftShader to test rendering results on all platforms without issue (Mac, Linux & Windows). When there are issues, CI will catch them and we can address the same on a case by case basis. There is no reason yet to make developers on Macs less confident in the test suite.

Reply:
@chinmaygarde : so before Skia gold is available for the engine repo, do you suggest us to submit a golden file to the engine repo for each OS (i.e., `screenshot_linux.png`, `screenshot_windows.png`, `screenshot_mac.png`)? I think that would be hard to maintain as anyone who updates the test need to run the test in 3 OSes to update the golden files...

Reply:
@chinmaygarde : oh, I see what you mean: upload only one golden file initially, and only disable non-Linux tests when CI report failures. I'm Ok with that :-) I guess that should work most of the time if we avoid anti-aliased geometries.

Reply:
Yeah, most results are stable. I hadn't considered `screenshot_<platform>.png` variants where the results did not match but that certainly something to consider. Bottom line, the value of the test suite is not derived from the fact that it successfully passes on all platforms (maybe by skipping some of the problematic stuff), but that it actually tests stuff on all platforms. Specifically in this case, you added a test for Fuchsia but the test would not have checked snapshotting on Fuchsia even if you were able to run the test suite there.  I am sure we will have to think harder about making sure we can test the engine consistently across all platforms but in this specific case, there isn't much to do :)

Reply:
My plan in this case was to get the test loaded, then once I get the shell_unittests working on Fuchsia (currently a high priority item), check and possibly generate a Fuchsia gold image and enable on Fuchsia, hence the TODO comment.